### DISCUSSION TITLE
Recommended method of Upgrading?

### DISCUSSION THREAD

**Discussion by chrish112** [PrivateBin/PrivateBin#1363] [Q&A]
### Describe the problem/question

Hi Everyone,
I have successfully been using PrivateBin and implemented it in my place of work but would like to know the best method of upgrading from 1.6.2 to the latest version. I have extensively adjusted the boostrap.php so that when you go to the website it looks different than the base version. Same with the config file and the data in the directory, am I to make a backup of these then drop in those files over the new ones? Re-do the ownership of the files and modify them? Everytime there's a new release?

### Did you use the FAQ section?

- [X] Yes, I have read [the FAQ](https://github.com/PrivateBin/PrivateBin/wiki/FAQ) and I found no solution/answer there.

### What you did?

This is a question

### What should happen

Any future options to basically grab the latest files, put them into a 2nd storage location for my Privatebin versions and then run a script that asks me to upgrade and then pull the edits from the old files and put them into the new?

### Server OS

Ubuntu 22 Server

### Webserver

Apache2

### PrivateBin version

1.7.2

### Issue reproducibility

No, I cannot reproduce it on https://privatebin.net.

**Comment by elrido** [comment]
I would recommend to rename your custom template to something different than bootstrap.php (not bootstrap5.php, though, maybe bootstrap-custom.php?). Then you can indeed just copy the files from the release archive over the existing installation. You will then want to copy the updated the SRI hashes (those long hex-strings in the script tag) from the new bootstrap.php onto your template or use some other method of regenerating them in your template.

Personally, I have come to like the convenience of using Linux container images, hosted on your container-runtime of choice (podman, docker, k8s, ...) and with data folder, configuration files and (in your case) template attached into the image. Hence my motivation to offer images for this project as well. We have two flavors ATM, one with a [traditional nginx + php-fpm stack](https://hub.docker.com/r/privatebin/nginx-fpm-alpine) and the other using the [newer nginx unit application server](https://hub.docker.com/r/privatebin/unit-alpine/). Both come in variants for the different backends and a chunkier all-in-one image that supports every backend (useful for migrations).

Some people use git as an upgrade mechanism. Checkout a specific release and use that same mechanism to switch from one release to the next. The configuration file and your custom template (if renamed) would just show up as two uncommitted files and not conflict any of the files from source control. We do try and support this as well, which is why we have vendored in the php-files from composer libraries and third party js-files.

**Comment by chrish112** [comment]
Hi @elrido ,
Thank you for responding. I'm still new to development but learned about breaking SRI tokens when I tried to tinker with the javascript files. At this time I'm still a little unsure of SRI tokens but I found an old comment of yours that linked back to this [github link](https://github.com/PrivateBin/PrivateBin/wiki/Development#subresource-integrity-for-javascript-resources). From the link I see the

As for the copying of the SRI hashes I believe I have found it within the bootstrap.php, it should look like this below?
`cript type="text/javascript" data-cfasync="false" src="js/zlib-1.3.1.js" integrity="sha512-Z90`
This is only one line but it looks like I'll need to copy over all the new strings or use something like a SRI Hash Generator? If that's the case the SRI hashes are all in the bootstrap? 

It does seem that your containers will be a better option but I'll probably need to sit down for a long time and really learn those. I started with Apache2 actually just a few months ago and started a couple sites, such as PrivateBin.

**Reply by elrido** [reply]
Yes, you should be able to copy over all the <script...> tags from the bootstrap.php template, if that is the one you copied. All the ones that have the integrity-attribute.

I wonder if having to do this on every upgrade is one of the reasons we don't see more templates getting used? The SRI-hashes are the same for all templates, and could be injected into them from a single source, like from (default) configuration values, so they could still be overridden in the configuration file, if really needed, but otherwise would always be up to date upon upgrade. I'll ponder on this.

**Reply by chrish112** [reply]
If the integrity attribute are the same for all the templates then I don't really need to inject the script into my edited bootstrap then? Maybe adjust the source to match the version of the file in the folders.

As for providing an update on my progress I'm kind of in a standstill again. I recognized that Ubuntu upgrade from 20 to 22 keeps the privatebin 1.6.2 version stable but if I make 2nd or 3rd virtual host folders for the Privatebin 1.7.3 and the third for my PrivateBin 1.7.3-Edited and tell apache to run those there's a problem. 

I thought it was Apache, and might still be right to a degree, because if I direct the virtual host to the non-edited 1.7.3 I'm greeted with the index.php information. I upgraded to the PHP 8.1, enabled the module, and now I'm able to see the 1.7.3 non edited webpage but redirecting back to using the 1.6.2 version it is blank. Same thing for the edited 1.7.3 as well. In the edited folder it should only be the edited bootstrap.php template that was changed.

I never really understand how Apache knows what to run in the folders that are created in order to produce a gui but assuming there's a file in Privatebin that tells what is supposed to be ran. Such as you have bootstrap5, how do  I tell apache to ignore bootstrap.php and use bootstrap5.php?

**Reply by rugk** [reply]
Well there are many resources on how a web server works, it's too much to explain here: https://duckduckgo.com/?t=ffab&q=how+does+apache+work&ia=web

About the template files though: You can ignore others, you don't need. You just configure the one you want (by file name) in the config file: https://github.com/PrivateBin/PrivateBin/wiki/Configuration#template

**Comment by rugk** [comment]
Another tip for upgrading templates: You may want to check the git history of this project of the template you've used as the basis. Just go to https://github.com/PrivateBin/PrivateBin/tree/master/tpl and select your template. Then click on history.

You can [also go to the latest release page on GitHub](https://github.com/PrivateBin/PrivateBin/releases/latest) and then click on Compare, go to "Files Changed" and find the template file there. [More information and alternative ways here](https://stackoverflow.com/questions/14500240/how-can-i-generate-a-diff-for-a-single-file-between-two-branches-in-github).

This may give you a hint on what to (potentially) adjust etc.
